mediatek: add support for Mercusys MR85X

Schneider-Azima opened the pull request:
This commit adds support for Mercusys MR85X router.

Device specification:
```
 - SoC: Mediatek MT7981b, Cortex-A53, 64-bit
 - RAM: 512MB
 - Flash: SPI NAND GigaDevice GD5F1GQ5UEYIGY (128 MB)
 - Ethernet: Realtek rtl8367s + 2.5Gbe Airoha en8811h
 - Ethernet: 1x2.5Gbe (WAN 2.5Gbps), 3xGbe (LAN 1Gbps, port0, port1, port2)
 - Wireless: 2.4GHz (802.11 b/g/n/ax)
 - Wireless: 5GHz (802.11 a/n/ac/ax)
 - LEDs: 1 amber and 1 green status LEDs, 4 green gpio-controlled LEDs
   on ethernet ports
 - Buttons: 2 (reset,wps)
 - Bootloader: Main U-Boot - U-Boot 2022.07-rc3. Additionally, both UBI
   slots contain "seconduboot" (also U-Boot 2022.07-rc3)
```

Installation (UART):
```
 - Place OpenWrt initramfs-kernel image on tftp server with IP 192.168.1.2
 - Attach UART, switch on the router and interrupt the boot process by
   pressing 'Ctrl-C'.
 - Set the uboot environment for startup.
   setenv tp_boot_idx 0; saveenv
   If the bootarg is set to boot from ubi1, also change it to ubi0.
 - Load and run OpenWrt initramfs image.
   setenv serverip 192.168.1.2; setenv ipaddr 192.168.1.1; tftpboot initramfs-kernel.bin; bootm
 - Browse IP 192.168.1.1, upload the 'sysupgrade' image and do upgrade.
```

Recovery:
```
 - Press Reset button and power on the router.
 - Navigate to U-Boot recovery web server (http://192.168.1.1/) and
   upload the OEM firmware.
```

Stock layout:
```
0x000000000000-0x000000200000 : "boot"
0x000000200000-0x000000300000 : "u-boot-env"
0x000000300000-0x000003500000 : "ubi0"
0x000003500000-0x000006700000 : "ubi1"
0x000006700000-0x000006f00000 : "userconfig"
0x000006f00000-0x000007300000 : "tp_data"
```

ubi0/ubi1 format:
U-Boot at boot checks that all volumes are in place:
```
+-------------------------------+
| Volume Name: uboot   Vol ID: 0|
| Volume Name: kernel  Vol ID: 1|
| Volume Name: rootfs  Vol ID: 2|
+-------------------------------+
```

MAC addresses:
```
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| label   | 00:FF:xx:xx:xx:EE | label     |
| WAN     | 00:FF:xx:xx:xx:EF | label+1   |
| LAN     | 00:FF:xx:xx:xx:EE | label     |
| WLAN 2g | 00:FF:xx:xx:xx:ED | label-1   |
| WLAN 5g | 00:FF:xx:xx:xx:EC | label-2   |
+---------+-------------------+-----------+
```
label MAC address was found in UBI partition "tp_data", file "default-mac".
OEM wireless eeprom is also there, file "MT7981_EEPROM.bin".

Schneider-Azima commented:
One confusing thing: this switch's id is 0x636700A0, and it supports up to 2.5G with cpu.

According to rtl8367b driver and rtl8365mb driver, this id corresponds to "rtl8367s" (rtl8367c family), which supports up to 1G as rgmii. The 2.5G model of rtl8367s, aka "rtl8367s-vb" (rtl8367d family), should have an id of 0x66420010.
#4178 #14804 

According to MR85X's GPL dts code, it seems to use "rtkgsw" (rtl8367c) driver. I tried, and it works.
https://www.mercusys.com/en/support/gpl-code?model=MR85X

I also tried rtl8367b driver and rtl8365mb driver. They can recognize the switch, but LAN don't work, no message interaction at all. If anyone knows how to make it work, please comment.

For rtl8367b driver, my dts node is as follow:
```
rtl8367s {
    compatible = "realtek,rtl8367b";
    realtek,extif = <7 1 0 1 1 1 1 1 1 2>;
    mii-bus = <&mdio>;
    phy-id = <29>;
};
```

For rtl8365mb driver, my dts node is as follow:
```
switch@1d {
    compatible = "mediatek,rtk-gsw";
    reg = <29>;
    reset-gpios = <&pio 39 GPIO_ACTIVE_LOW>;

    interrupt-controller {
        #interrupt-cells = <1>;
        interrupt-parent = <&pio>;
        interrupts = <38 IRQ_TYPE_LEVEL_HIGH>;
    };

    ports {
        #address-cells = <1>;
        #size-cells = <0>;

        port@0 {
            reg = <0>;
            label = "lan0";
        };

        port@1 {
            reg = <1>;
            label = "lan1";
        };

        port@2 {
            reg = <2>;
            label = "lan2";
        };

        port@6 {
            reg = <6>;
            label = "cpu";
            ethernet = <&gmac0>;
            phy-mode = "2500base-x";

            fixed-link {
                speed = <2500>;
                full-duplex;
                pause;
            };
        };
    };
};
```

musashino205 commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -859,6 +859,21 @@ endef
 $(eval $(call KernelPackage,switch-rtl8367b))
 
 
+define KernelPackage/switch-rtl8367s
+  SUBMENU:=$(NETWORK_DEVICES_MENU)
+  TITLE:=Realtek RTL8367S switch support
+  KCONFIG:=CONFIG_RTL8367S_GSW
+  FILES:=$(LINUX_DIR)/drivers/net/phy/rtk/rtl8367s_gsw.ko
+  AUTOLOAD:=$(call AutoLoad,43,rtl8367s_gsw,1)
+endef
+
+define KernelPackage/switch-rtl8367s/description
+ Realtek RTL8367S switch support
+endef
+
+$(eval $(call KernelPackage,switch-rtl8367s))
+
+
```
rtl8367s_gsw (rtl8367c) driver is not in the `target/linux/generic` directory and available only in mediatek target. So please move this entry to [`target/linux/mediatek/modules.mk`](https://github.com/openwrt/openwrt/blob/bbd95c0523174dc724a40b4d2483cadf7dc9907c/target/linux/mediatek/modules.mk).

Or, you can enable the `CONFIG_RTL8367S_GSW` symbol in the **config-6.xx** instead of adding that package.

https://github.com/openwrt/openwrt/blob/bbd95c0523174dc724a40b4d2483cadf7dc9907c/target/linux/mediatek/filogic/config-6.6#L434
https://github.com/openwrt/openwrt/blob/bbd95c0523174dc724a40b4d2483cadf7dc9907c/target/linux/mediatek/filogic/config-6.12#L455

Schneider-Azima replied:
Done.
I move it to `target/linux/mediatek/modules.mk`.
This switch can be recognized by rtl8367b/rtl8365mb/rtl8367s-gsw driver. To avoid future conflict, it's better not set in config-6.x directly.

dangowrt commented:
I'd very much prefer to get rid of that ugly driver altogether in favor of replacing it with the DSA driver. Different initvals will be needed for sure, and probably also other changes to the driver, but it would be so much better than keep using the old swconfig phy driver...

luizluca commented:
> Actually the whole RTL8365MB driver only support up to 1G. For RTL8367S switch, maybe EXT2 (extint 7) can work with it? I don't know. But EXT1 (extint 6) don't work.

RTL8365MB still does not support any other interface type but RGMII:

https://github.com/torvalds/linux/blob/master/drivers/net/dsa/realtek/rtl8365mb.c#L43C1-L44C1

> In other words, RTL8365MB driver does not fully support rtl8367s switch. To support 2.5G, most of this driver needs to be modified.
> 
> ```
> {
> 	.name = "RTL8367S",
> 	.chip_id = 0x6367,
> 	.chip_ver = 0x00A0,
> 	.extints = {
> 		{ 6, 1, PHY_INTF(SGMII) | PHY_INTF(HSGMII) },
> 		{ 7, 2, PHY_INTF(MII) | PHY_INTF(TMII) |
> 			PHY_INTF(RMII) | PHY_INTF(RGMII) },
> 	},
> 	.jam_table = rtl8365mb_init_jam_8365mb_vc,
> 	.jam_size = ARRAY_SIZE(rtl8365mb_init_jam_8365mb_vc),
> },
> ```

When I added these interfaces descriptions, I added (for completeness) both interfaces. However the driver didn't implement all those types as the two devs involved (me and Alvin Šipraga) didn't have access to devices that uses (H)SGMII. The actual information that the driver passes to the rest of the kernel is from rtl8365mb_phylink_get_caps(). Today it just checks for RGMII intefaces and sets it as capable and bails out if not:

https://github.com/torvalds/linux/blob/master/drivers/net/dsa/realtek/rtl8365mb.c#L1047C37-L1047C71

The driver is not responsible to warn if a mode is supported or not: just to list what it supports (get_caps). You might need to extend rtl8365mb_phylink_get_caps(), and the rest of the driver, to consider with interface is being configured and which types it supports. 

rtl8365mb_phylink_mac_config and rtl8365mb_phylink_mac_link_{up,down} are also only treating RGMII case as well as you'll need to adapt or creating a new rtl8365mb_ext_config_rgmii for your case (HSGMII/SGMII).

I would avoid adding more registers to the jam_table if HSGMII/SGMII needs some initialization. The jam_table is just an opaque bunch of registers values that should, one day, be gone. The rtl8367c is a good inspiration of what is needed to be configured to get your port up. This might help as well:

https://cdn.jsdelivr.net/gh/libc0607/Realtek_switch_hacking@files/Realtek_Unmanaged_Switch_ProgrammingGuide.pdf

> At this moment, I haven't seen any model that uses RTL8367S and supports 2.5G and is using RTL8365MB driver. If you insist to use RTL8365MB driver for 2.5G RTL8367S switch and make some change to the driver, do you have a candidate model for test? I suggest making this model the first one.

At this moment, no other dev has a device, like MR85X, to implement the HSGMII/SGMII interface.

dangowrt commented:
@hauke No, all (also most recent) MediaTek MACs support only MediaTek DSA special tags for PPE offloading. The work-around is to use Q-in-Q instead of the special tag, DSA has some helpers for that. Most modern DSA hardware has support for switching between a proprietary special tag and using Q-in-Q as tag format. The proprietary format has the advantage that all Q and Q-in-Q aka. af VLANs can be used, the Q-in-Q tag format has the advantage of being compatible with various offloading engines while limiting Q-in-Q on user ports.

Schneider-Azima commented:
> I added SGMII support for the rtl8365mb driver some time ago, but never finished it: https://patchwork.kernel.org/project/netdevbpf/list/?series=639580&state=* I also have some OpenWrt branches.

Thanks for your work. I tried it, IP protocol works, but TCP TX hardware checksum is wrong. If I trun off hardware checksum with `ethtool -K eth0 tx off`, network link seems OK. As eth0's mtu is 1508 (1500 and a tag?), I suppose this cause the error.

Now I remove the change to old phy switch driver, MR85X still works, with all LAN LEDs always on. Is it OK to merge?

luizluca commented:
> @hauke No, all (also most recent) MediaTek MACs support only MediaTek DSA special tags for PPE offloading. The work-around is to use Q-in-Q instead of the special tag, DSA has some helpers for that. Most modern DSA hardware has support for switching between a proprietary special tag and using Q-in-Q as tag format. The proprietary format has the advantage that all Q and Q-in-Q aka. af VLANs can be used, the Q-in-Q tag format has the advantage of being compatible with various offloading engines while limiting Q-in-Q on user ports.

Q-in-Q is a nice workaround. Losing Q-in-Q on the user ports is preferable to losing offload. The issue with Mediatek offload is that you cannot indicate where the IP packet starts, so it needs to parse the entire packet. If it doesn't expect a second layer of dot1Q, it won’t find the IP and TCP/UDP headers—similar to what happens with an alien (Realtek) CPU tag.

I assume that the DSA switch driver doesn’t need to explicitly support Q-in-Q, since it could consume the outer VLAN tag independently of what's encapsulated inside. Are these assumptions correct?

My test system uses a ramips/mt7620 with an rtl8367s. I'm not sure if Q-in-Q will work, but I can give it a try. Is there any supported device currently using it?

Schneider-Azima commented:
Changed `01_leds` script, use `ucidef_set_led_switch` to set LAN LED control. Now functionality seems all OK with rtl8367s_gsw driver.

Schneider-Azima commented:
luizluca@a0b554a This way test works with mt7981. But ethtool isn't aware of this. Maybe need some more code to expose the change of HW checksum to the stat interface?

```
root@OpenWrt:~# ethtool -k eth0 | grep checksum
rx-checksuming: on
tx-checksuming: on
	tx-checksum-ipv4: on
	tx-checksum-ip-generic: off [fixed]
	tx-checksum-ipv6: on
	tx-checksum-fcoe-crc: off [fixed]
	tx-checksum-sctp: off [fixed]
```

Forwarding offload still don't work. Seems @luizluca already know this.

@hauke 's HSGMII support seems unlikely to be accepted by kernel for quite a period of time. So may I open a new PR to support MR85X with DSA switch? There will be two OpenWrt kernel patches, one for HSGMII support, one for MTK HW checksum. Shall QinQ be completed one day and rtl8365mb enters a relatively usable state, simply delete/modify these two patches, and MR85X can still works without some new support commits.

I can help with test. MR85X uses ext1, works in HSGMII mode.

Schneider-Azima commented:
I made a new PR #19445 to support Mercusys MR85X with rtl8365mb DSA switch driver.

This PR uses the old rtl8367s_gsw swconfig driver. Functionality seems all OK.
If it is clearly decided that no more new device will rely on this old driver, please let me know. And I will close this one. Thanks.

hauke commented:
I did not follow up on the upstream submission. If you want you can continue the effort. I do not know how to handle the firmware needed for the HSGMII part of the switch. It is part of a binary with a GPL header.

hauke commented on target/linux/mediatek/modules.mk:
```diff
@@ -37,3 +37,14 @@ define KernelPackage/iio-mt6577-auxadc
+  KCONFIG:=CONFIG_RTL8367S_GSW
+  FILES:= \
+	$(LINUX_DIR)/drivers/net/phy/rtk/rtl8367s_gsw.ko
+  AUTOLOAD:=$(call AutoLoad,43,rtl8367s_gsw,1)
```
does Autoprobe work too?
```suggestion
  AUTOLOAD:=$(call AutoProbe,rtl8367s_gsw,1)
```

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1333,6 +1333,19 @@ define Device/mercusys_mr80x-v3
+  DEVICE_MODEL := MR85X
+  DEVICE_DTS := mt7981b-mercusys-mr85x
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware kmod-phy-airoha-en8811h kmod-swconfig swconfig kmod-switch-rtl8367s
```
```suggestion
  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware kmod-phy-airoha-en8811h swconfig kmod-switch-rtl8367s
```
Not needed with extra dependency.

Schneider-Azima commented:
@dangowrt @hauke @luizluca @namiltd 
Hi, may I kindly suggest that we merge this PR as a first support for this device? And I will change #19445 to a "convert to DSA".

As we continue to wait for the DSA driver to be completed, PRs need to be constantly updated to handle merge conflicts. "convert" only needs to modify dts and network initialization scripts, and things will be much simpler.

luizluca commented:
> @dangowrt @hauke @luizluca @namiltd Hi, may I kindly suggest that we merge this PR as a first support for this device? And I will change #19445 to a "convert to DSA".
> 
> As we continue to wait for the DSA driver to be completed, PRs need to be constantly updated to handle merge conflicts. "convert" only needs to modify dts and network initialization scripts, and things will be much simpler.

Whether the device is using swconfig or DSA should not block the merge. I just don't know how devs feels about  rtl8367s_gsw. It was born as a target specific module and it looks like it was kept as a low profile. Being a generic code, independent from arch/target, it should have landed in the linux/generic, not mediatek.

Schneider-Azima committed:
mediatek: build rtl8367s_gsw as a kernel module

The RTL8367S_GSW config for filogic target is not set by default.
This switch model can be recognized by multiple drivers. Configure it as an optional module instead of default set in config.

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>
Link: https://github.com/openwrt/openwrt/pull/19187
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Schneider-Azima committed:
mediatek: add support for Mercusys MR85X

This commit adds support for Mercusys MR85X router.

Device specification:
 - SoC: Mediatek MT7981b, Cortex-A53, 64-bit
 - RAM: 512MB
 - Flash: SPI NAND GigaDevice GD5F1GQ5UEYIGY (128 MB)
 - Ethernet: Realtek rtl8367s + 2.5Gbe Airoha en8811h
 - Ethernet: 1x2.5Gbe (WAN 2.5Gbps), 3xGbe (LAN 1Gbps, port0, port1, port2)
 - Wireless: 2.4GHz (802.11 b/g/n/ax)
 - Wireless: 5GHz (802.11 a/n/ac/ax)
 - LEDs: 1 amber and 1 green status LEDs, 4 green gpio-controlled LEDs
   on ethernet ports
 - Buttons: 2 (reset,wps)
 - Bootloader: Main U-Boot - U-Boot 2022.07-rc3. Additionally, both UBI
   slots contain "seconduboot" (also U-Boot 2022.07-rc3)

Installation (UART):
 - Place OpenWrt initramfs-kernel image on tftp server with IP 192.168.1.2
 - Attach UART, switch on the router and interrupt the boot process by
   pressing 'Ctrl-C'.
 - Set the uboot environment for startup.
   setenv tp_boot_idx 0; saveenv
   If the bootarg is set to boot from ubi1, also change it to ubi0.
 - Load and run OpenWrt initramfs image.
   setenv serverip 192.168.1.2; setenv ipaddr 192.168.1.1; tftpboot initramfs-kernel.bin; bootm
 - Browse IP 192.168.1.1, upload the 'sysupgrade' image and do upgrade.

Recovery:
 - Press Reset button and power on the router.
 - Navigate to U-Boot recovery web server (http://192.168.1.1/) and
   upload the OEM firmware.

Stock layout:
0x000000000000-0x000000200000 : "boot"
0x000000200000-0x000000300000 : "u-boot-env"
0x000000300000-0x000003500000 : "ubi0"
0x000003500000-0x000006700000 : "ubi1"
0x000006700000-0x000006f00000 : "userconfig"
0x000006f00000-0x000007300000 : "tp_data"

ubi0/ubi1 format:
U-Boot at boot checks that all volumes are in place:
+-------------------------------+
| Volume Name: uboot   Vol ID: 0|
| Volume Name: kernel  Vol ID: 1|
| Volume Name: rootfs  Vol ID: 2|
+-------------------------------+

MAC addresses:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| label   | 00:FF:xx:xx:xx:EE | label     |
| WAN     | 00:FF:xx:xx:xx:EF | label+1   |
| LAN     | 00:FF:xx:xx:xx:EE | label     |
| WLAN 2g | 00:FF:xx:xx:xx:ED | label-1   |
| WLAN 5g | 00:FF:xx:xx:xx:EC | label-2   |
+---------+-------------------+-----------+
label MAC address was found in UBI partition "tp_data", file "default-mac".
OEM wireless eeprom is also there, file "MT7981_EEPROM.bin".

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>
Link: https://github.com/openwrt/openwrt/pull/19187
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
